server/src/main/scala/org/http4s/server/middleware/ConcurrentRequests.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.server.middleware
+
+import cats.data._
+import cats.effect._
+import cats.effect.concurrent._
+import cats.implicits._
+import org.http4s._
+import org.http4s.server._
+
+/** Middlewares for tracking the quantity of concurrent requests.
+  *
+  * These are generalized middlewares and can be used to implement metrics,
+  * logging, max concurrent requests, etc.
+  *
+  * @note The concurrent request count is decremented on the completion of the
+  *       Response body, or in the event of any error, and is guaranteed to
+  *       only occur once.
+  */
+object ConcurrentRequests {
+
+  /** Run a side effect each time the concurrent request count increments and
+    * decrements.
+    *
+    * @note Each side effect is given the current number of concurrent
+    *       requests as an argument.
+    *
+    * @note `onIncrement` should never be < 1 and `onDecrement` should never
+    *       be value < 0.
+    *
+    * @note This is the same as [[#route]], but allows for the inner and outer
+    *       effect types to differ.
+    */
+  def route_[F[_]: Sync, G[_]: Sync](
+      onIncrement: Long => G[Unit],
+      onDecrement: Long => G[Unit]
+  ): F[ContextMiddleware[G, Long]] =
+    Ref
+      .in[F, G, Long](0L)
+      .map(ref =>
+        BracketRequestResponse.bracketRequestResponseRoutes[G, Long](
+          ref.updateAndGet(_ + 1L).flatTap(onIncrement)
+        )(Function.const(ref.updateAndGet(_ - 1L).flatMap(onDecrement))))
+
+  /** Run a side effect each time the concurrent request count increments and
+    * decrements.
+    *
+    * @note Each side effect is given the current number of concurrent
+    *       requests as an argument.
+    *
+    * @note `onIncrement` should never be < 1 and `onDecrement` should never
+    *       be value < 0.
+    */
+  def route[F[_]: Sync](
+      onIncrement: Long => F[Unit],
+      onDecrement: Long => F[Unit]
+  ): F[ContextMiddleware[F, Long]] =
+    route_[F, F](onIncrement, onDecrement)
+
+  /** As [[#apply]], but runs the same effect on increment and decrement of the concurrent request count. */
+  def onChangeRoute[F[_]: Sync](onChange: Long => F[Unit]): F[ContextMiddleware[F, Long]] =
+    route[F](onChange, onChange)
+
+  /** Run a side effect each time the concurrent request count increments and
+    * decrements.
+    *
+    * @note Each side effect is given the current number of concurrent
+    *       requests as an argument.
+    *
+    * @note `onIncrement` should never be < 1 and `onDecrement` should never
+    *       be value < 0.
+    *
+    * @note This is the same as [[#app]], but allows for the inner and outer
+    *       effect types to differ.
+    */
+  def app_[F[_]: Sync, G[_]: Sync](
```

Comment:
I think we called this `2` instead of `_` when we did similar in `Http4sDsl`.

Reply:
@rossabaker I'll update. `_` is my default `'` from idioms Haskell. I never know what to name things.

Reply:
updated

Reply:
I think of the `_` as being unit, as in `traverse_`, but that's possibly just Cats looking for an `'` as well.